openquake/hazardlib/gsim/kale_2015.py

```diff
@@ -0,0 +1,385 @@
+        """
+        See :meth:`superclass method
+        <.base.GroundShakingIntensityModel.get_mean_and_stddevs>`
+        for spec of input and result values.
+
+        Implement equation 1, page 970.
+        """
+        # compute median PGA on rock, needed to compute non-linear site
+        # amplification
+        C_pga = self.COEFFS[PGA()]
+        median_pga = np.exp(
+            self._compute_mean(C_pga, rup.mag, dists, rup.rake)
+        )
+
+        # compute full mean value by adding nonlinear site amplification terms
+        C = self.COEFFS[imt]
+        mean = (self._compute_mean(C, rup.mag, dists, rup.rake) +
+                self._compute_non_linear_term(C, median_pga, sites))
+
+        stddevs = self._get_stddevs(C, rup, sites.vs30.shape, stddev_types)
+
+        return mean, stddevs
+
+    def _get_stddevs(self, C, rup, shape, stddev_types):
+        """
+        Return standard deviations as defined in p. 971.
+        """
+        weight = self._compute_weight_std(C, rup.mag)
+        std_intra = weight * C["sd1"] * np.ones(shape)
+        std_inter = weight * C["sd2"] * np.ones(shape)
+        
+        stddevs = []
+        for stddev_type in stddev_types:
+            assert stddev_type in self.DEFINED_FOR_STANDARD_DEVIATION_TYPES
+            if stddev_type == const.StdDev.TOTAL:
+                stddevs.append(np.sqrt(std_intra ** 2. + std_inter ** 2.))
+            elif stddev_type == const.StdDev.INTRA_EVENT:
+                stddevs.append(std_intra)
+            elif stddev_type == const.StdDev.INTER_EVENT:
+                stddevs.append(std_inter)
+        return stddevs
+
+    def _compute_weight_std(self, C, mag):
+        """
+        Common part of equations 8 and 9, page 971.
+        """
+        if mag < 6.0:
+            return C['a1']
+        elif mag >= 6.0 and mag < 6.5:
+            return C['a1'] + (C['a2'] - C['a1']) * ((mag - 6.0) / 0.5)
+        else:
+            return C['a2']
+
+    def _compute_magnitude_scaling_term(self, C, mag):
+        """
+        Compute and return magnitude scaling term in equation 2,
+        page 970.
+        """
+        c1 = self.CONSTS['c1']
+        if mag <= c1:
+            return C['b1'] + C['b2'] * (mag - c1) + C['b3'] * (8.5 - mag) ** 2
+        else:
+            return C['b1'] + C['b7'] * (mag - c1) + C['b3'] * (8.5 - mag) ** 2
+
+    def _compute_geometric_decay_term(self, C, mag, dists):
+        """
+        Compute and return geometric decay term in equation 3,
+        page 970.
+        """
+        c1 = self.CONSTS['c1']
+        return (
+            (C['b4'] + C['b5'] * (mag - c1)) *
+            np.log(np.sqrt(dists.rjb ** 2.0 + C['b6'] ** 2.0))
+        )
+
+    def _compute_faulting_style_term(self, C, rake):
+        """
+        Compute and return style-of-faulting term in equation 4,
+        page 970.
+        """
+        Fn = float(rake > -135.0 and rake < -45.0)
+        Fr = float(rake > 45.0 and rake < 135.0)
+

```

Comment:
If this is used in other places I have no objections. It is just a convention I have never seen before.